From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Sun, 12 Jun 2022 10:57:51 -0500
Subject: [PATCH] Thunderbolt configuration file


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 2b3786560d8fb93b9899ebc1e0a7c96a808268d0..731b074f6a9ee0dd3129fe177fe1dfd8a4cc54d8 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2233,6 +2233,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Plazma end
 
+        // Thunderbolt start - Thunderbolt Configuration API
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getThunderboltConfig() {
+            throw new UnsupportedOperationException("Not implemented yet.");
+        }
+        // Thunderbolt end
+
         /**
          * Sends the component to the player
          *
