From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Sun, 12 Jun 2022 10:57:51 -0500
Subject: [PATCH] Volt configuration file


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 2b3786560d8fb93b9899ebc1e0a7c96a808268d0..70685d2d6fe869e1338e74518a252adf70099c4f 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2233,6 +2233,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Plazma end
 
+        // Volt start - Volt Configuration API
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getVoltConfig() {
+            throw new UnsupportedOperationException("Not implemented yet.");
+        }
+        // Volt end
+
         /**
          * Sends the component to the player
          *
