From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Fri, 10 May 2024 12:50:20 +0900
Subject: [PATCH] Add basic configuration parts


diff --git a/src/main/java/org/plazmamc/thunderbolt/configurations/GlobalConfiguration.java b/src/main/java/org/plazmamc/thunderbolt/configurations/GlobalConfiguration.java
index fbac06d6c70fd6250f16a6bdbea21873ff5e1d2a..1b1d08f00716551a26a5cee1133d2ce77fc8d1b7 100644
--- a/src/main/java/org/plazmamc/thunderbolt/configurations/GlobalConfiguration.java
+++ b/src/main/java/org/plazmamc/thunderbolt/configurations/GlobalConfiguration.java
@@ -25,4 +25,22 @@ public class GlobalConfiguration extends ConfigurationPart {
     @Setting(Configuration.VERSION_FIELD)
     int version = VERSION;
 
+    public NeedsReload needsReload;
+    public class NeedsReload extends ConfigurationPart {
+
+
+        @PostProcess
+        public void post() {
+            boolean changed = ThunderboltConfigurations.GLOBAL_LOADED;
+
+            if (changed) {
+                ThunderboltConfigurations.LOGGER.warn("Configuration settings that require a restart have been changed."
+                        + "The settings will take effect at the next restart.");
+                return;
+            }
+
+        }
+
+    }
+
 }
diff --git a/src/main/java/org/plazmamc/thunderbolt/configurations/ThunderboltConfigurations.java b/src/main/java/org/plazmamc/thunderbolt/configurations/ThunderboltConfigurations.java
index 2b6653148b6a3586373efffa7a74caf2dd5e560c..219a2ec97a69abdcf7f018c053b1d014ee4b42e2 100644
--- a/src/main/java/org/plazmamc/thunderbolt/configurations/ThunderboltConfigurations.java
+++ b/src/main/java/org/plazmamc/thunderbolt/configurations/ThunderboltConfigurations.java
@@ -160,7 +160,7 @@ public class ThunderboltConfigurations extends Configurations<GlobalConfiguratio
 
     @Override
     protected WorldConfigurations getWorldConfiguration(ServerLevel level) {
-        throw new UnsupportedOperationException("TODO");
+        return level.thunderboltConfig();
     }
 
     @Override
